File: examples/validate_embeddings.py
#!/usr/bin/env python3
"""
Validation script to compare Python ColBERT embeddings with Rust.

Run this after running the Rust validate_embeddings example and compare outputs.

Usage: python validate_embeddings.py
"""

import numpy as np
from fastembed import LateInteractionTextEmbedding
from fastembed.postprocess import Muvera


def maxsim(query_emb: np.ndarray, doc_emb: np.ndarray) -> float:
    """Compute ColBERT MaxSim score."""
    # query_emb: (q_tokens, dim), doc_emb: (d_tokens, dim)
    # For each query token, find max similarity with any doc token
    similarities = query_emb @ doc_emb.T  # (q_tokens, d_tokens)
    return float(similarities.max(axis=1).sum())


def main():
    # Fixed test inputs - MUST match Rust script exactly
    documents = [
        "The quick brown fox jumps over the lazy dog.",
        "Machine learning is a subset of artificial intelligence.",
    ]
    queries = [
        "What animal jumps?",
        "What is ML?",
    ]

    print("=== Python ColBERT + MUVERA Validation ===\n")

    # Initialize model
    model = LateInteractionTextEmbedding("colbert-ir/colbertv2.0")
    print(f"Model dimension: {model.embedding_size}")

    # Embed documents
    print("\n--- Document Embeddings ---")
    doc_embeddings = list(model.embed(documents))
    for i, emb in enumerate(doc_embeddings):
        print(f"Doc {i} shape: {emb.shape}")
        if emb.size > 0:
            print(f"  First token, first 5 values: {emb[0, :5].tolist()}")
            norm = np.linalg.norm(emb[0])
            print(f"  First token L2 norm: {norm:.6f}")
            print(f"  First token sum: {emb[0].sum():.6f}")

    # Embed queries
    print("\n--- Query Embeddings ---")
    query_embeddings = list(model.query_embed(queries))
    for i, emb in enumerate(query_embeddings):
        print(f"Query {i} shape: {emb.shape}")
        if emb.size > 0:
            print(f"  First token, first 5 values: {emb[0, :5].tolist()}")
            norm = np.linalg.norm(emb[0])
            print(f"  First token L2 norm: {norm:.6f}")
            print(f"  First token sum: {emb[0].sum():.6f}")

    # Test MaxSim scores (brute force ColBERT)
    print("\n--- MaxSim Scores (Query x Doc) ---")
    for qi, q_emb in enumerate(query_embeddings):
        for di, d_emb in enumerate(doc_embeddings):
            score = maxsim(q_emb, d_emb)
            print(f"Query {qi} x Doc {di}: {score:.4f}")

    # Test MUVERA
    print("\n--- MUVERA FDEs ---")
    muvera = Muvera(
        dim=model.embedding_size,
        k_sim=5,
        dim_proj=16,
        r_reps=20,
        random_seed=42,
    )
    print(f"MUVERA embedding size: {muvera.embedding_size}")

    doc_fdes = [muvera.process_document(emb) for emb in doc_embeddings]
    query_fdes = [muvera.process_query(emb) for emb in query_embeddings]

    for i, fde in enumerate(doc_fdes):
        print(f"Doc {i} FDE first 5: {fde[:5].tolist()}")
        print(f"  FDE L2 norm: {np.linalg.norm(fde):.6f}")

    for i, fde in enumerate(query_fdes):
        print(f"Query {i} FDE first 5: {fde[:5].tolist()}")
        print(f"  FDE L2 norm: {np.linalg.norm(fde):.6f}")

    # MUVERA dot product scores
    print("\n--- MUVERA Dot Product Scores (Query x Doc) ---")
    for qi, q_fde in enumerate(query_fdes):
        for di, d_fde in enumerate(doc_fdes):
            score = float(np.dot(q_fde, d_fde))
            print(f"Query {qi} x Doc {di}: {score:.4f}")


if __name__ == "__main__":
    main()

File: examples/validate_embeddings.rs
//! Validation script to compare Rust ColBERT embeddings with Python
//!
//! Run this, then run validate_embeddings.py and compare outputs.
//!
//! Usage: cargo run --example validate_embeddings --features muvera

use anyhow::Result;
use fastembed::{LateInteractionInitOptions, LateInteractionModel, LateInteractionTextEmbedding, Muvera};

fn main() -> Result<()> {
    // Fixed test inputs - MUST match Python script exactly
    let documents = vec![
        "The quick brown fox jumps over the lazy dog.",
        "Machine learning is a subset of artificial intelligence.",
    ];
    let queries = vec![
        "What animal jumps?",
        "What is ML?",
    ];

    println!("=== Rust ColBERT + MUVERA Validation ===\n");

    // Initialize model
    let mut model = LateInteractionTextEmbedding::try_new(
        LateInteractionInitOptions::new(LateInteractionModel::ColBERTV2)
    )?;

    println!("Model dimension: {}", model.dim());

    // Embed documents
    println!("\n--- Document Embeddings ---");
    let doc_embeddings = model.embed(&documents, None)?;
    for (i, emb) in doc_embeddings.iter().enumerate() {
        println!("Doc {} shape: ({}, {})", i, emb.len(), if emb.is_empty() { 0 } else { emb[0].len() });
        if !emb.is_empty() && !emb[0].is_empty() {
            // Print first token's first 5 values
            println!("  First token, first 5 values: {:?}", &emb[0][..5.min(emb[0].len())]);
            // Print L2 norm of first token (should be ~1.0)
            let norm: f32 = emb[0].iter().map(|x| x * x).sum::<f32>().sqrt();
            println!("  First token L2 norm: {:.6}", norm);
            // Print sum of first token
            let sum: f32 = emb[0].iter().sum();
            println!("  First token sum: {:.6}", sum);
        }
    }

    // Embed queries
    println!("\n--- Query Embeddings ---");
    let query_embeddings = model.query_embed(&queries, None)?;
    for (i, emb) in query_embeddings.iter().enumerate() {
        println!("Query {} shape: ({}, {})", i, emb.len(), if emb.is_empty() { 0 } else { emb[0].len() });
        if !emb.is_empty() && !emb[0].is_empty() {
            println!("  First token, first 5 values: {:?}", &emb[0][..5.min(emb[0].len())]);
            let norm: f32 = emb[0].iter().map(|x| x * x).sum::<f32>().sqrt();
            println!("  First token L2 norm: {:.6}", norm);
            let sum: f32 = emb[0].iter().sum();
            println!("  First token sum: {:.6}", sum);
        }
    }

    // Test MaxSim scores (brute force ColBERT)
    println!("\n--- MaxSim Scores (Query x Doc) ---");
    for (qi, q_emb) in query_embeddings.iter().enumerate() {
        for (di, d_emb) in doc_embeddings.iter().enumerate() {
            let score = maxsim(q_emb, d_emb);
            println!("Query {} x Doc {}: {:.4}", qi, di, score);
        }
    }

    // Test MUVERA
    println!("\n--- MUVERA FDEs ---");
    let muvera = Muvera::from_late_interaction_model(
        &model,
        Some(5),   // k_sim
        Some(16),  // dim_proj
        Some(20),  // r_reps
        Some(42),  // seed
    )?;
    println!("MUVERA embedding size: {}", muvera.embedding_size());

    let doc_fdes: Vec<Vec<f32>> = doc_embeddings.iter()
        .map(|emb| muvera.process_document(emb))
        .collect();

    let query_fdes: Vec<Vec<f32>> = query_embeddings.iter()
        .map(|emb| muvera.process_query(emb))
        .collect();

    for (i, fde) in doc_fdes.iter().enumerate() {
        println!("Doc {} FDE first 5: {:?}", i, &fde[..5]);
        println!("  FDE L2 norm: {:.6}", fde.iter().map(|x| x * x).sum::<f32>().sqrt());
    }

    for (i, fde) in query_fdes.iter().enumerate() {
        println!("Query {} FDE first 5: {:?}", i, &fde[..5]);
        println!("  FDE L2 norm: {:.6}", fde.iter().map(|x| x * x).sum::<f32>().sqrt());
    }

    // MUVERA dot product scores
    println!("\n--- MUVERA Dot Product Scores (Query x Doc) ---");
    for (qi, q_fde) in query_fdes.iter().enumerate() {
        for (di, d_fde) in doc_fdes.iter().enumerate() {
            let score: f32 = q_fde.iter().zip(d_fde.iter()).map(|(a, b)| a * b).sum();
            println!("Query {} x Doc {}: {:.4}", qi, di, score);
        }
    }

    Ok(())
}

/// Compute ColBERT MaxSim score
fn maxsim(query_emb: &[Vec<f32>], doc_emb: &[Vec<f32>]) -> f32 {
    query_emb
        .iter()
        .map(|q| {
            doc_emb
                .iter()
                .map(|d| q.iter().zip(d.iter()).map(|(a, b)| a * b).sum::<f32>())
                .fold(f32::NEG_INFINITY, f32::max)
        })
        .sum()
}
